pkg/acsengine/defaults-apiserver.go

```diff
@@ -0,0 +1,107 @@
+
+	// Data Encryption at REST configuration
+	if helpers.IsTrueBoolPointer(o.KubernetesConfig.EnableDataEncryptionAtRest) {
+		staticLinuxAPIServerConfig["--experimental-encryption-provider-config"] = "/etc/kubernetes/encryption-config.yaml"
+	}
+
+	// Aggregated API configuration
+	if o.KubernetesConfig.EnableAggregatedAPIs {
+		staticLinuxAPIServerConfig["--requestheader-client-ca-file"] = "/etc/kubernetes/certs/proxy-ca.crt"
+		staticLinuxAPIServerConfig["--proxy-client-cert-file"] = "/etc/kubernetes/certs/proxy.crt"
+		staticLinuxAPIServerConfig["--proxy-client-key-file"] = "/etc/kubernetes/certs/proxy.key"
+		staticLinuxAPIServerConfig["--requestheader-allowed-names"] = ""
+		staticLinuxAPIServerConfig["--requestheader-extra-headers-prefix"] = "X-Remote-Extra-"
+		staticLinuxAPIServerConfig["--requestheader-group-headers"] = "X-Remote-Group"
+		staticLinuxAPIServerConfig["--requestheader-username-headers"] = "X-Remote-User"
+	}
+
+	// Enable cloudprovider if we're not using cloud controller manager
+	if !helpers.IsTrueBoolPointer(o.KubernetesConfig.UseCloudControllerManager) {
+		staticLinuxAPIServerConfig["--cloud-provider"] = "azure"
+		staticLinuxAPIServerConfig["--cloud-config"] = "/etc/kubernetes/azure.json"
+	}
+
+	// AAD configuration
+	if cs.Properties.HasAadProfile() {
+		staticLinuxAPIServerConfig["--oidc-username-claim"] = "oid"
+		staticLinuxAPIServerConfig["--oidc-client-id"] = "spn:" + cs.Properties.AADProfile.ServerAppID
+		issuerHost := "sts.windows.net"
+		if GetCloudTargetEnv(cs.Location) == "AzureChinaCloud" {
+			issuerHost = "sts.chinacloudapi.cn"
+		}
+		staticLinuxAPIServerConfig["--oidc-issuer-url"] = "https://" + issuerHost + "/"
+	}
+
+	staticWindowsAPIServerConfig := make(map[string]string)
+	for key, val := range staticLinuxAPIServerConfig {
+		staticWindowsAPIServerConfig[key] = val
+	}
+	// Windows apiserver config overrides
+	// TODO placeholder for specific config overrides for Windows clusters
+
+	// Default apiserver config
+	defaultAPIServerConfig := map[string]string{}
+
+	// If no user-configurable apiserver config values exists, use the defaults
+	if o.KubernetesConfig.APIServerConfig == nil {
+		o.KubernetesConfig.APIServerConfig = defaultAPIServerConfig
+	} else {
+		for key, val := range defaultAPIServerConfig {
+			// If we don't have a user-configurable apiserver config for each option
+			if _, ok := o.KubernetesConfig.APIServerConfig[key]; !ok {
+				// then assign the default value
+				o.KubernetesConfig.APIServerConfig[key] = val
+			}
+		}
+	}
+
+	// We don't support user-configurable values for the following,
+	// so any of the value assignments below will override user-provided values
+	var overrideAPIServerConfig map[string]string
+	if cs.Properties.HasWindows() {
+		overrideAPIServerConfig = staticWindowsAPIServerConfig
+	} else {
+		overrideAPIServerConfig = staticLinuxAPIServerConfig
+	}
+	for key, val := range overrideAPIServerConfig {
+		o.KubernetesConfig.APIServerConfig[key] = val
```

Comment:
Some argument like "--admission-control" and "--authorization-mode" takes a list of features, and probably should allow feature addition (as long as static config is a subset of the user input config), instead of override. It can be done in a later change if needed.

Reply:
I agree this would be a useful addition to this surface area. Feature gates also are special, in that they essentially contain key/vals themselves. Yeah, I'd prefer a follow-up PR for that.